enhancements/windows-containers/bring-your-own-windows-host.md

```diff
@@ -0,0 +1,171 @@
+  of all components installed by WMCO.
+
+### Non-Goals
+
+* Installing the container runtime in the Windows instance
+* Handle upgrading of the Windows operating system
+* OpenShift Builds
+* OpenShift DeploymentConfigs
+
+## Proposal
+
+The cluster admin will create a ConfigMap called _windows-instances_ in the
+_openshift-windows-machine-config-operator_ namespace. WMCO will be configured
+to watch for this special ConfigMap, whose data section will have the following
+layout:
+```yaml
+kind: ConfigMap
+apiVersion: v1
+metadata:
+  name: windows-instances
+  namespace: openshift-windows-machine-config-operator
+data:
+  instance.1: |-
+    username=Administrator
+    ip-address=10.1.42.1
+  instance.2: |-
+    username=core
+    ip-address=10.1.42.2
+```
+WMCO will configure or de-configure instances based on the contents of the data
+section and will use the `cloud-private-key` to communicate with them.
+
+### User Stories
+
+#### Story 1
+As an OpenShift cluster administrator I would like to add an existing Windows
+instance to the cluster as a worker node.
+
+#### Story 2
+As an OpenShift cluster administrator I would like to remove an existing Windows
+instance from the cluster as a worker node.
+
+#### Story 3
+As an OpenShift cluster administrator I would like to have the Kubernetes
+components on Windows nodes upgraded when new Kubernetes versions or patch
+versions are released.
+
+### Risks and Mitigations
+
+The same risk and mitigation that apply to Windows nodes added using MachineSets
+to an IPI cluster applies here.
+
+## Design Details
+
+Changes to the _windows-instances_ ConfigMap will cause WMCO to compare the
+list of instances with the set of existing Windows nodes. The comparison will
+be based on the instance and node's IP addresses. WMCO will then react in the
+following manner:
+* If a new entry is found i.e. the list of nodes does not contain one that map
+  to the instances' IP address, the instance will be configured by WMCO. The
+  configuration process will be identical to configuring instances added using
+  a MachineSet in an IPI cluster.
+* If the list of nodes has IP addresses that do not map to any in the list of
+  instances, WMCO will perform an uninstall operation on the node. This will
+  involve:
+  * Drain the node
+  * Stop and remove
+    * Monitoring and logging services
+    * kube-proxy service
+    * hybrid-overlay-node service
+    * kubelet service
+  * Delete the payload binaries
+* If the list of nodes has IP addresses that do not map to any in the list of
+  instances, and WMCO is unable to access the instance, it will assume that
+  the node in question is not managed by it.
+
+The status of these operations will be reflected by events in the
+_openshift-windows-machine-config-operator_ namespace and will be associated
+with the _windows-instances_ ConfigMap.
+
+The upgrade process for BYOH nodes will be identical to upgrading Windows nodes
+added using a MachineSet in an IPI cluster.
```

Comment:
How? The IPI upgrade process involves deleting the host. We cant do that in this case, we'd need to upgrade in place

Reply:
That's what I thought too. At a high level the upgrade should be
- Updated version of WMCO notices that there is a version mismatch
- Should follow the delete logic which was mentioned https://github.com/aravindhp/enhancements/pull/3/files#diff-4589867d0ff2cbc590c57e37b0a3c1eb5b111e69c4b47cc87658b4abdf5508a3R110
- Reconfigure the Windows VM since it notices that there is no node object associated with instances present in the configmap

To be clear the trigger for this controller should not only be from configmap but for also for node objects.

Reply:
I agree that I should clarify this a bit more, though I don't full under the flow written by @ravisantoshgudimetla 

> Reconfigure the Windows VM since it notices that there is no node object associated with instances present in the configmap

This will happen independent of the upgrade process. Upgrade process is only for instances mentioned in the ConfigMap and that have a corresponding node object. If there is an instance in the ConfigMap without a node object that is akin to adding a new instance.

>To be clear the trigger for this controller should not only be from configmap but for also for node objects.

@ravisantoshgudimetla, not sure what you mean by this. We trigger upgrades based on the version annotation. 

Or are you saying that we need to map changes to the node object to the ConfigMap like we do for Machines? Yes, we need to do that. However I am not sure if that needs to be called here.

Reply:
This should be deleted as its now conflicting with the section you added in the `Upgrade` section

Reply:
> Or are you saying that we need to map changes to the node object to the ConfigMap like we do for Machines? Yes, we need to do that. However I am not sure if that needs to be called here.

That is correct. However there is some terminology confusion. Controller should be watching for node objects as well. I think it is better to call that out.

Reply:
I like the idea of having configmap for upgrades but I am fine with the simpler solution we discussed in the meeting. The actual route we decide on can be added as an update to the current upgrade enhancement proposal later.